### DISCUSSION TITLE
Consuming updates?

### DISCUSSION THREAD

**Discussion by escott-irt** [RomanIakovlev/timeshape#121] [General]
Sorry if I missed it somewhere in the documentation, but I am wondering how updates to the IANA TZ database (which consequently trigger updates to timezone-boundary-builder) would be absorbed?

Would I need to rebuild my software with a new version of timeshape? Or is there some way to consume an update at runtime (similar to how the JVM has a tzupdater.jar Timezone Updater Tool to update the data set).

**Comment by RomanIakovlev** [comment]
Timeshape relies on Java runtime handling the `tzdb` updates. If your `tzdb` in Java runtime is up to date (either by upgrading to the latest Java version or running `tzupdater`), Timeshape will pick up the changes when initialized, e.g. when one of the `initialize` overloads is called, e.g. [this one](https://github.com/RomanIakovlev/timeshape/blob/master/core/src/main/java/net/iakovlev/timeshape/TimeZoneEngine.java#L152). As far as I understand, this requires re-start of your Java runtime, or at least re-initializing the `TimeZoneEngine` class.

Under the hood Timeshape just keeps a mapping between time zone polygons and string representation of timezone IDs. The timezone IDs are picked from the source data in timezone-boundary-builder every time a new release of Timeshape is built. During initialization Timeshape tries to create the relevant instances of `java.time.ZoneId` class [here](https://github.com/RomanIakovlev/timeshape/blob/master/core/src/main/java/net/iakovlev/timeshape/Index.java#L163), using `java.time.ZoneId::of` method, which only succeeds if your Java runtime knows about the given timezone ID.

Hope this helps. Please feel free to follow up if any part of this explanation is not clear! A PR adding an explanation of this process to the readme/docs would be welcome as well! 😃

**Comment by escott-irt** [comment]
Very helpful, thank you.

The latter part - when the time zone polygon to time zone ID map is updated, and there is a new build of Timeshape consuming that update - I'd need to rebuild my software depending on that new build, correct? There's no way to consume the update in the source data at run-time?

It sounds like an inconsistency between the `tzdb` and the Timeshape version may also cause problems, if a timezone ID exists in one and not the other.

**Comment by RomanIakovlev** [comment]
> There's no way to consume the update in the source data at run-time?

There is such way, actually. It's not super-prominent in the readme, but if you look at the last paragraph [here](https://github.com/RomanIakovlev/timeshape#using-the-library) you can read this:

> Data can also be read from an external file, see overloads of TimeZoneEngine.initialize that accept TarArchiveInputStream. The file format should be same as produced by running builder sbt project (see [here](https://github.com/RomanIakovlev/timeshape/blob/master/doc/Architecture.md#builder)). It is responsibility of caller to close input stream passed to TimeZoneEngine.initialize. 

Using this approach you can build the release data separately and re-initialize the `TimeZoneEngine` with new data. This way you can get fresh time zone data without re-deploying your main program. This feature is a community contribution, I haven't personally used that myself, so give it a try and feel free to report any feedback.

P.S. I wonder how popular this use case is, to update the Timeshape data without re-building the main program. Maybe there should be a more convenient way to do this? @escott-irt if you feel like the existing solution could be improved, I'd gladly discuss how this could be done and can provide implementation guidance, if you'd be interested in taking on this.

**Comment by escott-irt** [comment]
Ah! Thanks for explaining that. I had noticed that mentioned in the readme but not grasped what it meant. 

I would speculate that updating the `timeshape` data without re-building would be a popular request - the data set is evolving and in some industries the software release cycle can be expensive.

The ability to initialize based on an external `tar.zstd` archive is sufficient; an ability to fetch pre-built `tar.zstd` archives would be a convenient addition. Assuming it is important that Java's `tzdb` and the `timeshape` data set are consistent then some kind of tool to fetch the appropriate `tar.zstd` archive for the current JVM's `tzdb` version might be ideal.

We are just researching options right now and not yet making use of `timeshape`, so I'm not in a position to make more specific suggestions or contributions, but if I get to that point I will share.
